### ISSUE TITLE
float8: Revisit approach to keeping amax/scale buffers in fp32

### ISSUE THREAD

**Issue by vkuzo** [pytorch/ao#561]
from @awgu 

**Overview**
The goal is to have the amax, amax history, and scale tensors be (1) always kept in fp32 and (2) included in the state dict.

The issue is that (a) if these tensors are registered as buffers, then they are affected by `_apply()` methods like `.half()` or `.to(torch.bfloat16)` or similar and that (b) if these tensors are plain attributes (not buffers), then they are not included in the state dict. As such, valid approaches could either keep them as buffers but handle the dtype or could keep them as plain attributes and handle the state dict.

**Current Approach**
The current approach involves overriding `_apply()` to re-cast these buffers to fp32 in case the `_apply()` function changed their dtype.

https://github.com/pytorch-labs/float8_experimental/blob/d0c67607f8de58390184e1a621ed69af92433a8e/float8_experimental/float8_linear.py#L151-L165

**Discussion**
My issue with this approach is that re-casting to fp32 does not prevent precision loss from the preceding downcast and that it only prevents dtype change from `nn.Module._apply()` methods (e.g. `module.to(torch.bfloat16)`), not `Tensor` methods (`buf.to(torch.bfloat16)`).

For example, if `_apply()` to downcast is called after loading the state dict, then we may lose valuable precision from the buffers, quietly affecting numerics.

One alternative would be to wrap the buffers in a lightweight tensor subclass that overrides `.to()` to make any dtype cast a no-op, enforcing that the buffer preserves its original fp32 dtype. However, we would need to check that this subclass works with the rest of the infrastructure (e.g. `torch.compile()` and FSDP).

Another alternative would be to keep the tensors as plain attributes and override state dict (e.g. using the provided hooks) to include them in the state dict. If users do not expect these to appear in `named_buffers()`, then this could be a reasonable idea.

copied from https://github.com/pytorch-labs/float8_experimental/issues/156
